src/python/pants/backend/debian/rules.py

```diff
@@ -1,64 +1,89 @@
+from pants.core.util_rules.system_binaries import BinaryPathRequest, BinaryPaths, TarBinary
+from pants.engine.fs import CreateDigest, DigestEntries, FileEntry
+from pants.engine.internals.native_engine import Digest
 from pants.engine.internals.selectors import Get
-from pants.engine.process import Process, ProcessResult
+from pants.engine.process import Process, ProcessCacheScope, ProcessResult
 from pants.engine.rules import collect_rules, rule
+from pants.engine.target import HydratedSources, HydrateSourcesRequest
 from pants.engine.unions import UnionRule
 from pants.util.logging import LogLevel
 
 
 @dataclass(frozen=True)
 class DebianPackageFieldSet(PackageFieldSet):
-    required_fields = (DebianControlFile, DebianInstallPrefix, DebianPackageDependencies)
+    required_fields = (DebianSources, DebianInstallPrefix, DebianPackageDependencies)
 
-    control: DebianControlFile
+    sources_dir: DebianSources
     install_prefix: DebianInstallPrefix
     packages: DebianPackageDependencies
     output_path: OutputPathField
 
 
-@rule(level=LogLevel.DEBUG)
-async def package_debian_package(field_set: DebianPackageFieldSet) -> BuiltPackage:
+@rule(level=LogLevel.INFO)
+async def package_debian_package(
+    field_set: DebianPackageFieldSet, tar_binary_path: TarBinary
+) -> BuiltPackage:
     dpkg_deb_path = await Get(
         BinaryPaths,
         BinaryPathRequest(
-            binary_name="touch",
+            binary_name="dpkg-deb",
             search_path=["/usr/bin"],
         ),
     )
     if not dpkg_deb_path.first_path:
-        raise OSError("Could not find the `touch` program on search paths ")
+        raise OSError(f"Could not find the `{dpkg_deb_path.binary_name}` program on search paths ")
+
+    hydrated_sources = await Get(HydratedSources, HydrateSourcesRequest(field_set.sources_dir))
 
-    output_filename = field_set.output_path.value_or_default(file_ending="deb")
+    # since all the sources are coming only from a single directory, it is
+    # safe to pick an arbitrary file and get its root directory name
+    sources_directory_name = PurePath(hydrated_sources.snapshot.files[0]).parts[0]
 
-    # TODO(alexey): add Debian packaging logic
     result = await Get(
         ProcessResult,
         Process(
             argv=(
-                "touch",
-                output_filename,
+                dpkg_deb_path.first_path.path,
+                "--build",
+                sources_directory_name,
             ),
             description="Create a Debian package from the produced packages.",
-            output_files=(output_filename,),
+            input_digest=hydrated_sources.snapshot.digest,
+            # dpkg-deb produces a file with the same name as the input directory
+            output_files=(f"{sources_directory_name}.deb",),
+            env={"PATH": str(PurePath(tar_binary_path.path).parent)},
+            cache_scope=ProcessCacheScope.PER_SESSION,
```

Comment:
Remind me why we need this? It doesn't surprise me, but I'm not certain. It'd be good to also add that explanation as a comment.

Reply:
I think I had it set during development to make sure it's always running the code as I make changes (instead of just picking up the cached artifacts). I've tested packaging multiple times in different configurations without the cache scope set and it works fine. I've removed this.

Done!